homeassistant/components/apple_tv/media_player.py

```diff
@@ -48,49 +49,53 @@
 )
 
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Set up the Apple TV platform."""
-    if not discovery_info:
-        return
-
-    # Manage entity cache for service handler
-    if DATA_ENTITIES not in hass.data:
-        hass.data[DATA_ENTITIES] = []
-
-    name = discovery_info[CONF_NAME]
-    host = discovery_info[CONF_HOST]
-    atv = hass.data[DATA_APPLE_TV][host][ATTR_ATV]
-    power = hass.data[DATA_APPLE_TV][host][ATTR_POWER]
-    entity = AppleTvDevice(atv, name, power)
-
-    @callback
-    def on_hass_stop(event):
-        """Stop push updates when hass stops."""
-        atv.push_updater.stop()
-
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
-
-    if entity not in hass.data[DATA_ENTITIES]:
-        hass.data[DATA_ENTITIES].append(entity)
-
-    async_add_entities([entity])
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Load Apple TV media player based on a config entry."""
+    name = config_entry.data[CONF_NAME]
+    manager = hass.data[DOMAIN][config_entry.unique_id]
+    async_add_entities([AppleTvDevice(name, config_entry.unique_id, manager)])
 
 
 class AppleTvDevice(MediaPlayerEntity):
     """Representation of an Apple TV device."""
 
-    def __init__(self, atv, name, power):
+    def __init__(self, name, identifier, manager):
         """Initialize the Apple TV device."""
-        self.atv = atv
+        self.atv = None
         self._name = name
+        self._identifier = identifier
         self._playing = None
-        self._power = power
-        self._power.listeners.append(self)
-        self.atv.push_updater.listener = self
+        self._manager = manager
 
     async def async_added_to_hass(self):
         """Handle when an entity is about to be added to Home Assistant."""
-        self._power.init()
+        self._manager.listeners.append(self)
+        await self._manager.init()
```

Comment:
Why do we wait to init the manager until the entity is added?

Reply:
My intention is to be sure that the entities exists and receive the connected/disconnected callbacks from the manager. Connecting earlier would create a race condition.

Reply:
Can we store the connection state in the manager? Then the entity could just access the current connection state of the manager when it has been added. The entity callbacks for updating connection state can access this saved state too.

Reply:
The problem is that these listeners are used to hook up the listener used for playback metadata (which is not handled by the manager in any way). So the media player could potentially miss metadata updates.

Reply:
Ok. A better way of waiting for the platforms is to create a task that finishes the integration setup from `async_setup_entry` in the init module. In that task we await the `asyncio.gather` of forward platforms. After that we await `manager.init`. See eg the ozw integration.

https://github.com/home-assistant/core/blob/e4d29bf3ec39cddd54a7f30a6c16879f73a2cf2c/homeassistant/components/ozw/__init__.py#L228-L241

Reply:
That sounds perfect. In theory I can fetch the metadata without the push interface, but I have a regression with`DMAP` (Apple TV 3 and earlier) if I do. It will trigger a new request which will interfere with the push update handler internally in `pyatv` because of how `DMAP` works. You can't do both at the same time. It will work but might generate unwanted warnings in the log. Once I fix this I do it the way you suggested first.